rusoto/core/src/v2.rs

```diff
@@ -0,0 +1,79 @@
+//! TODO
+
+use futures::{IntoFuture, Future};
+
+use crate::Client;
+use crate::Region;
+use crate::{RusotoResult, RusotoFuture};
+use crate::error::RusotoError;
+use crate::request::HttpResponse;
+use crate::signature::SignedRequest;
+
+/// TODO
+pub trait ServiceRequest {
+    /// TODO
+    type Output;
+    /// TODO
+    type Error;
+
+    /// TODO
+    fn dispatch(self, region: &Region, dispatcher: &impl Dispatcher) -> RusotoFuture<Self::Output, Self::Error>;
+}
+
+/// TODO
+pub struct Request<T> {
+    input: T,
+    region: Region,
+    client: Client,
+}
+
+impl<T: ServiceRequest> Request<T> {
+    /// TODO
+    pub fn new(input: T, region: Region, client: Client) -> Self {
+        Request { input, region, client }
+    }
+
+    /// TODO
+    pub fn send(self) -> RusotoFuture<T::Output, T::Error> {
+        self.input.dispatch(&self.region, &self.client)
+    }
+
+    /// TODO
+    pub fn sync(self) -> RusotoResult<T::Output, T::Error>
```

Comment:
Another point of clarification—is the intent of the v2 api (which I like a lot!) to _also_ align with the upcoming stabilization of async/await? If so, I think it'd be feasible to drop the `.sync()` and the `RusotoFuture` and replace that with a [`BoxFuture`](https://rust-lang-nursery.github.io/futures-api-docs/0.3.0-alpha.16/futures/future/type.BoxFuture.html)

(I think taking things one step at a time is very reasonable, however, as this is a bigger change.)

Reply:
It would definitely be nice to have a good story around async/await here. I haven't played around much with it in nightly yet, but the `IntoFuture` trait on `Request` should work out of the box with `await`, such that you could do something like `s3.get_item(req).await?`.

I'm thinking we might want to keep the `.sync()` compatbility layer around for a little longer and deprecate it once the ecosystem has finished moving to async/await?

Reply:
> It would definitely be nice to have a good story around async/await here. I haven't played around much with it in nightly yet, but iiuc the IntoFuture trait on Request should make it work out of the box with await?

Yep, I think you're right about that. That's a miss on my part!

> I'm thinking we might want to keep the .sync() compatbility layer around for a little longer and deprecate it once the ecosystem has finished moving to async/await?

Yeah, I can go either way with this. With https://github.com/awslabs/aws-lambda-rust-runtime, I've been [prototyping a Lambda runtime](https://github.com/davidbarsky/proto-lambda) that uses `std::future::Future` + `async/await` + [runtime](https://github.com/davidbarsky/proto-lambda). Most notably, backwards compatibility isn't a constraint I'm holding myself to. In doing so, I've found that the `async/await` feature is pretty robust, minus a few questionable diagnostics. I think I projected my own reasoning/assumptions onto your work here, for which I apologize!

If you're curious, I expect that the prototype lambda interface would look a lot like:

```rust
use http::{Request, Response};

type Err = Box<dyn std::error::Error + Send + Sync + 'static>;

#[lambda]
#[runtime::main]
async fn main(req: Request<Vec<u8>>) -> Result<Response<Vec<u8>, Err> {
    match req.uri().path() {
        "/" => index(req).await?,
        "/hello" => hello(req).await?,
        _ => not_found(req)?,
    }
}
```